File: C++/LeetCode/0034.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> result(2, -1);

        int start = 0, end = nums.size() - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (nums[mid] >= target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        if (start >= nums.size() || nums[start] != target) {
            return result;
        }

        result[0] = start;

        end = nums.size() - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (nums[mid] <= target) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        result[1] = end;

        return result;
    }
};

File: JAVA/LeetCode/0034.java
class Solution {
    static int first(int[] nums, int target) {
        int n = nums.length;
        int s = 0;
        int e = n - 1;

        int ans = -1;

        while(s <= e) {
            int mid = s+ (e-s)/2;
            if(nums[mid] == target) {
                ans = mid;
                e = mid - 1;
            }else if(nums[mid] > target) {
                e = mid - 1;
            }else{
                s = mid + 1;
            }
        }
        return ans;
    }

    static int last(int[] nums, int target) {
        int n = nums.length;
        int s = 0;
        int e = n - 1;

        int ans = -1;
        
        while(s <= e) {
            int mid = s+ (e-s)/2;
            if(nums[mid] == target) {
                ans = mid;
                s = mid + 1;
            }else if(nums[mid] > target) {
                e = mid - 1;
            }else{
                s = mid + 1;
            }
        }
        return ans;
    }
    public int[] searchRange(int[] nums, int target) {
        return new int[] {first(nums, target), last(nums, target)};
    }
}